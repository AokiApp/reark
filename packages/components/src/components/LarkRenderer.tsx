import React, { useContext } from "react";
import { LarkApiContext } from "../contexts/LarkApiContext";
import { BlockStoreProvider } from "../contexts/BlockStoreContext";
import { BlockComponent } from "./BlockComponent";

export const LarkRenderer: React.FC = () => {
  const { blocks } = useContext(LarkApiContext);

  if (!blocks || blocks.length === 0) return null;

  const rootBlock =
    blocks.find((block) => block.block_type === 1 && !block.parent_id) ||
    blocks[0];

  if (!rootBlock) return null;

  return (
    <BlockStoreProvider items={blocks}>
      <BlockComponent blockId={rootBlock.block_id} />
    </BlockStoreProvider>
  );
};
