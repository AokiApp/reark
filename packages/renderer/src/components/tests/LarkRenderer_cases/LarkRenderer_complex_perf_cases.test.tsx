import { renderWithVRT } from "../test-utils/renderWithVRT";
import { LarkRenderer } from "../../LarkRenderer";

// LarkRenderer 複合ケース: 複数ブロック混在
describe("LarkRenderer 複合ケース: 複数ブロック混在", () => {
  it("text, heading, bullet, table など複数ブロックを同時にレンダリングできる", async () => {
    const pageBlock = {
      block_id: "page-mix",
      block_type: 1,
      children: ["text-mix", "heading-mix", "bullet-mix", "table-mix"],
      parent_id: "",
    };
    const textBlock = {
      block_id: "text-mix",
      block_type: 2,
      parent_id: "page-mix",
      text: {
        elements: [
          {
            text_run: {
              content: "複合テスト用テキスト",
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: "left", folded: false },
      },
      children: [],
    };
    const headingBlock = {
      block_id: "heading-mix",
      block_type: 3,
      parent_id: "page-mix",
      heading1: {
        elements: [
          {
            text_run: {
              content: "見出し複合",
              text_element_style: { bold: true },
            },
          },
        ],
        style: { align: 1, folded: false },
      },
      children: [],
    };
    const bulletBlock = {
      block_id: "bullet-mix",
      block_type: 12,
      parent_id: "page-mix",
      bullet: {
        elements: [
          {
            text_run: {
              content: "リスト複合",
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: 1, folded: false },
      },
      children: [],
    };
    const tableBlock = {
      block_id: "table-mix",
      block_type: 18,
      parent_id: "page-mix",
      table: {
        cells: [
          [{ text: "セルA" }, { text: "セルB" }],
          [{ text: "セルC" }, { text: "セルD" }],
        ],
        property: {},
      },
      children: [],
    };

    const blocks = [
      pageBlock,
      textBlock,
      headingBlock,
      bulletBlock,
      tableBlock,
    ];

    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});

// LarkRenderer パフォーマンス: 1000個のtextブロック
describe("LarkRenderer パフォーマンス: 1000個のtextブロック", () => {
  it("大量のblocksでもレンダリングできる", async () => {
    const blocks = Array.from({ length: 1000 }).map((_, i) => ({
      block_id: `text-${i}`,
      block_type: 2,
      parent_id: "",
      text: {
        elements: [
          {
            text_run: {
              content: `テキスト${i}`,
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: "left", folded: false },
      },
      children: [],
    }));
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});

// LarkRenderer 入れ子構造: 10段ネストのbulletリスト
describe("LarkRenderer 入れ子構造: 10段ネストのbulletリスト", () => {
  it("10段ネストのbulletリストをレンダリングできる", async () => {
    const blocks = [];
    let parentId = "";
    for (let i = 0; i < 10; i++) {
      const block = {
        block_id: `bullet-nest-${i}`,
        block_type: 12,
        parent_id: parentId,
        bullet: {
          elements: [
            {
              text_run: {
                content: `ネスト${i}`,
                text_element_style: { bold: false },
              },
            },
          ],
          style: { align: "left", folded: false },
        },
        children: i < 9 ? [`bullet-nest-${i + 1}`] : [],
      };
      blocks.push(block);
      parentId = `bullet-nest-${i}`;
    }
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});
