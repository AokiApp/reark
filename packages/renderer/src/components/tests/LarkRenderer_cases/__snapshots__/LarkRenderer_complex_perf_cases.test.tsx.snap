// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LarkRenderer パフォーマンス: 1000個のtextブロック 大量のblocksでもレンダリングできる 1`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-text-block"
      style="text-align: left;"
    >
      <span
        class="reark-text"
      >
        テキスト0
      </span>
    </div>
  </div>
</div>
`;

exports[`LarkRenderer 入れ子構造: 10段ネストのbulletリスト 10段ネストのbulletリストをレンダリングできる 1`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-unordered-list"
    >
      <div
        class="reark-text-block"
        style="text-align: left;"
      >
        <span
          class="reark-text"
        >
          ネスト0
        </span>
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-unordered-list"
        >
          <div
            class="reark-text-block"
            style="text-align: left;"
          >
            <span
              class="reark-text"
            >
              ネスト1
            </span>
          </div>
          <div
            class="reark-block"
          >
            <div
              class="reark-unordered-list"
            >
              <div
                class="reark-text-block"
                style="text-align: left;"
              >
                <span
                  class="reark-text"
                >
                  ネスト2
                </span>
              </div>
              <div
                class="reark-block"
              >
                <div
                  class="reark-unordered-list"
                >
                  <div
                    class="reark-text-block"
                    style="text-align: left;"
                  >
                    <span
                      class="reark-text"
                    >
                      ネスト3
                    </span>
                  </div>
                  <div
                    class="reark-block"
                  >
                    <div
                      class="reark-unordered-list"
                    >
                      <div
                        class="reark-text-block"
                        style="text-align: left;"
                      >
                        <span
                          class="reark-text"
                        >
                          ネスト4
                        </span>
                      </div>
                      <div
                        class="reark-block"
                      >
                        <div
                          class="reark-unordered-list"
                        >
                          <div
                            class="reark-text-block"
                            style="text-align: left;"
                          >
                            <span
                              class="reark-text"
                            >
                              ネスト5
                            </span>
                          </div>
                          <div
                            class="reark-block"
                          >
                            <div
                              class="reark-unordered-list"
                            >
                              <div
                                class="reark-text-block"
                                style="text-align: left;"
                              >
                                <span
                                  class="reark-text"
                                >
                                  ネスト6
                                </span>
                              </div>
                              <div
                                class="reark-block"
                              >
                                <div
                                  class="reark-unordered-list"
                                >
                                  <div
                                    class="reark-text-block"
                                    style="text-align: left;"
                                  >
                                    <span
                                      class="reark-text"
                                    >
                                      ネスト7
                                    </span>
                                  </div>
                                  <div
                                    class="reark-block"
                                  >
                                    <div
                                      class="reark-unordered-list"
                                    >
                                      <div
                                        class="reark-text-block"
                                        style="text-align: left;"
                                      >
                                        <span
                                          class="reark-text"
                                        >
                                          ネスト8
                                        </span>
                                      </div>
                                      <div
                                        class="reark-block"
                                      >
                                        <div
                                          class="reark-unordered-list"
                                        >
                                          <div
                                            class="reark-text-block"
                                            style="text-align: left;"
                                          >
                                            <span
                                              class="reark-text"
                                            >
                                              ネスト9
                                            </span>
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LarkRenderer 複合ケース: 複数ブロック混在 text, heading, bullet, table など複数ブロックを同時にレンダリングできる 1`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-page"
    >
      <div
        class="reark-page-title"
      >
        <div
          class="reark-text-block"
          style="text-align: left;"
        />
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-text-block"
          style="text-align: left;"
        >
          <span
            class="reark-text"
          >
            複合テスト用テキスト
          </span>
        </div>
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-heading reark-heading--1"
        >
          <div
            class="reark-text-block"
            style="text-align: left;"
          >
            <span
              class="reark-text--bold reark-text"
            >
              見出し複合
            </span>
          </div>
        </div>
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-unordered-list"
        >
          <div
            class="reark-text-block"
            style="text-align: left;"
          >
            <span
              class="reark-text"
            >
              リスト複合
            </span>
          </div>
        </div>
      </div>
      <div
        class="reark-block"
      >
        <div
          class="reark-unsupported-block"
        >
          <strong>
            Unsupported block type: 
            18
          </strong>
        </div>
      </div>
    </div>
  </div>
</div>
`;
